walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,115 @@
+    if (request.getDuration() <= TimeRange.WHOLE_DAY.duration()){
+        Collection<TimeRange> mandatoryTimes = findAvailableMeetingTimes(events, request, true);
+        Collection<TimeRange> optionalTimes =  findAvailableMeetingTimes(events,request, false);
+        Collection<String> mandatoryAttendees = request.getAttendees();
+        Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+        if(optionalAttendees.size()== 0){
+            return mandatoryTimes;
+        }
+        if(mandatoryAttendees.size()== 0){
+            return optionalTimes;
+        }
+        Collection<TimeRange> commonTimes = findCommonTimes(mandatoryTimes,optionalTimes, request);
+        if(commonTimes.size() == 0){
+            return mandatoryTimes;
+        }
+        else{
+            return commonTimes;
+        }
+    
+    }
+    else {
+      return Collections.emptyList();
+    }
   }
-}
+  
+  //for mandatoryOrOptional, true means mandatory attendees and false means optional attendees
+   public Collection<TimeRange> findAvailableMeetingTimes(Collection<Event> events, MeetingRequest request, boolean mandatoryOrOptional){
+      Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>(Arrays.asList(TimeRange.WHOLE_DAY));
+      Collection<String> meetingAttendees;
+      if (mandatoryOrOptional == true){
+          meetingAttendees = request.getAttendees();
+      }
+      else{
+          meetingAttendees = request.getOptionalAttendees();
+      }
+      for (Event event : events){
+        Set<String> eventAttendees = event.getAttendees();
+        if (Collections.disjoint(meetingAttendees, eventAttendees) ==  false){ 
+        ArrayList<TimeRange> toRemove = new ArrayList<TimeRange>();
+          ArrayList<TimeRange> toAdd= new ArrayList<TimeRange>();
+          for(TimeRange freeslot : meetingTimes){
+            if (freeslot.overlaps(event.getWhen())){
+              if (event.getWhen().start()<=freeslot.end() && event.getWhen().start()>=freeslot.start()){
+                if (freeslot.start() != event.getWhen().start()){
+                  toAdd.add(TimeRange.fromStartEnd(freeslot.start(), event.getWhen().start(),false));
+                }
+              }
+              if (event.getWhen().end()<= freeslot.end() && event.getWhen().end()>= freeslot.start()){
+                if(event.getWhen().end() != freeslot.end()){
+                  toAdd.add(TimeRange.fromStartEnd(event.getWhen().end(), freeslot.end(),false));
+                }
+              }
+              toRemove.add(freeslot);
+            }
+          }
+          meetingTimes.removeAll(toRemove);
+          meetingTimes.addAll(toAdd);
+        }
+      }
+    ArrayList<TimeRange> durationTooShort = new ArrayList<TimeRange>();
+    for (TimeRange freeslot : meetingTimes){
+        if (freeslot.duration()< request.getDuration()){
+            durationTooShort.add(freeslot);
+        }
+    }
+    meetingTimes.removeAll(durationTooShort);
```

Comment:
Wouldn't it make sense to remove the too short meeting times before going through all of them?

Reply:
I moved inside the loop, so that after each iteration when the new available times are generated, the too short meetings are removed before the next iteration.